### PR TITLE
Rop builder

### PR DESCRIPTION
**Pull request by mexx**
I added a simple builder for ROP and converted some simple functions in `Simpilier` module to use the builder.

@forki @theimowski Please review
related to #464

### CODE REVIEW

File: src/Paket.Core/Simplifier.fs [hunk 1]
```diff
@@ -34,60 +34,50 @@ let removePackage(packageName, indirectPackages, fileName, interactive) =
...
-            referencesFiles
-            |> List.map (fun refFile -> simplifyReferencesFile(refFile, flatLookup, interactive))
-            |> Rop.collect
+    let referencesFiles' =
+        referencesFiles
+        |> List.map (fun refFile -> simplifyReferencesFile(refFile, flatLookup, interactive))
+        |> Rop.collect
 
-        let projects = List.zip projectFiles <!> referencesFiles'
+    let! projects = List.zip projectFiles <!> referencesFiles'
```
Comment (theimowski) [inline src/Paket.Core/Simplifier.fs]: ``referencesFiles'`` could be also bound with ``let!``, so that the ``<!>`` operator is not necessary here

**Comment by theimowski** [comment]
Looks good to me
Is it a good practice (in general) to maintain both custom operators and a custom builder for the same concept?

File: src/Paket.Core/Simplifier.fs (hunk 1 above)
Comment (mexx) [inline src/Paket.Core/Simplifier.fs]: you're right, I didn't notice that one.

**Comment by mexx** [comment]
I would go only for builder as for me it's more convenient to read the code.

**Comment by theimowski** [comment]
Ok I think I can live with only builder, @forki U have a opinion?

**Comment by forki** [comment]
then we should opt for builder...

2015-01-07 11:40 GMT+01:00 Tomasz Heimowski <notifications@github.com>:

**Comment by theimowski** [comment]
Ok I'll switch to the builder with new PRs

**Comment by isaacabraham** [comment]
I think we should have both a builder and custom operators options out there - oftentimes the builder will be more useful but occasionally e.g. if you just have one or two items in the pipeline - then operators are useful.

**Comment by vasily-kirichenko** [comment]
Why have you guys invented this "rop" thing?.. Isn't it the Either monad which is already implemented in FSharpx and ExtCore (+ lots of combinators and combined monads like AsyncChoice, ProtectedState etc.)?

**Comment by mexx** [comment]
@vasily-kirichenko You're right, it's the Either monad.
For me it has some learn effect, as I can dive into creating own computational expression builders.
Nevertheless we could replace the own implementation by referencing either of given alternatives.
@forki and @theimowski  Thoughts?

**Comment by forki** [comment]
I'm pretty happy with "domain specific" wording. Reusing either might be a
good idea, but only as implementation detail.

**Comment by theimowski** [comment]
I think the key point here is that we apply the technique, and whether we use custom or depend on another infrastructure is not that important

**Comment by mexx** [comment]
@theimowski well spoken

**Comment by vasily-kirichenko** [comment]
Two things: 1. code reuse 2. well known monads/names.

**Comment by isaacabraham** [comment]
@vasily-kirichenko makes a good point. To be fair I went through the pain of creating my own Either monad a while back, it was a good learning experience, but I think it might be worth using a tried-and-tested one that will be more "standard".

**Comment by forki** [comment]
The point of Rop is that it's more specific than Either monad. Of course it's an instance of an Either (and we can reuse implementation), but it gives meaning to really abstract Choice1Of2 and Choice2of2. It also comes with a nice metaphor which can make it easier to grasp the concept (we will see if this is true). 
So in one sense it's not the same as either - it's ab application of either.

**Pull request merged**